rdflib/plugins/serializers/patch.py

```diff
@@ -0,0 +1,95 @@
+from __future__ import annotations
+
+import warnings
+from typing import IO, Optional
+from uuid import uuid4
+
+from rdflib import Dataset, Graph
+from rdflib.plugins.serializers.nquads import _nq_row
+from rdflib.plugins.serializers.nt import _nt_row
+from rdflib.serializer import Serializer
+
+add_remove_methods = {"add": "A", "remove": "D"}
+
+
+class PatchSerializer(Serializer):
+    """
+    Creates an RDF patch file to add and remove triples/quads.
+    Can either:
+    - Create an add or delete patch for a single Dataset.
+    - Create a patch to represent the difference between two Datasets.
+    """
+
+    def __init__(
+        self,
+        store: Dataset,
+    ):
+        self.store = store
+        super().__init__(store)
+
+    def serialize(
+        self,
+        stream: IO[bytes],
+        base: Optional[str] = None,
+        encoding: Optional[str] = None,
+        operation: Optional[str] = None,
+        target: Optional[Graph] = None,
+        header_id: Optional[str] = None,
+        header_prev: Optional[str] = None,
+    ):
+        if not header_id:
+            header_id = f"uuid:{uuid4()}"
+        encoding = self.encoding
+        if base is not None:
+            warnings.warn("PatchSerializer does not support base.")
+        if encoding is not None and encoding.lower() != self.encoding.lower():
+            warnings.warn(
+                "PatchSerializer does not use custom encoding. "
+                f"Given encoding was: {encoding}"
+            )
+
+        def write_header():
+            stream.write(f"H id <{header_id}> .\n".encode(encoding, "replace"))
```

Comment:
It's consistent with the jsonld, nquads, and rdf xml serializers. I would think the default behaviour on encode would be better: `The default is 'strict' meaning that encoding errors raise a UnicodeEncodeError`. It would be good to know if there was a specific reason for the use of "replace" - @nicholascar @ashleysommer coincidentally you're on the commits for those serializers (a few years ago, I realise) - you wouldn't happen to recall or know why "replace" was used?

Reply:
I don't remember the exact reason. There was a bunch of stuff messed up and in a broken state after the python2->python3 `unicode`->`str` migration, I believe the solution was to pepper a bunch of ".encode()" method calls everywhere required, but we didn't want that to start causing code to break that was previously working, so using "replace" was a way of silently fixing it.
I agree however it is the wrong solution.
I've been doing some background work on bringing up the standard of the `stream` interface used by parsers and serializers, its another part that needs a big breaking change to fix, as part of RDFLib8.

Reply:
Discussed w/ Ashley  & Nick leaving making any encoding changes in a separate PR to generally address the issue.